test/unit/matrix/test_tile.cpp

```diff
@@ -341,3 +339,300 @@ TYPED_TEST(TileTest, CreateBuffer) {
   EXPECT_EQ(tile.size().rows(), data_blocksize(tile_data));
   EXPECT_EQ(tile.ld(), data_stride(tile_data));
 }
+
+template <class T, Device D>
+auto createTileAndPtrChecker(TileElementSize size, SizeType ld) {
+  memory::MemoryView<T, D> memory_view(ld * size.cols());
+  auto memory_view2 = memory_view;
+  Tile<T, D> tile(size, std::move(memory_view), ld);
+  // construct a second tile referencing the same memory for testing pointers
+  Tile<T, D> tile2(size, std::move(memory_view2), ld);
+  auto tile_ptr = [tile2 = std::move(tile2)](const TileElementIndex& index) { return tile2.ptr(index); };
+  CHECK_TILE_PTR(tile_ptr, tile);
+  return std::make_tuple(std::move(tile), std::move(tile_ptr));
+}
+
+template <class T, Device D>
+auto createTileChain() {
+  using T0 = std::remove_const_t<T>;
+  // set up tile chain
+  hpx::lcos::local::promise<Tile<T0, D>> tile_p;
+  hpx::future<Tile<T0, D>> tmp_tile_f = tile_p.get_future();
+  hpx::lcos::local::promise<Tile<T0, D>> next_tile_p;
+  hpx::future<Tile<T0, D>> next_tile_f = next_tile_p.get_future();
+
+  hpx::future<Tile<T, D>> tile_f =
+      tmp_tile_f.then(hpx::launch::sync,
+                      hpx::util::unwrapping([p = std::move(next_tile_p)](auto tile) mutable {
+                        tile.setPromise(std::move(p));
+                        return Tile<T, D>(std::move(tile));
+                      }));
+
+  return std::make_tuple(std::move(tile_p), std::move(tile_f), std::move(next_tile_f));
+}
+
+template <class F, class T>
+void checkSubtile(F&& ptr, T&& tile, SubTileSpec spec) {
+  auto subtile_ptr = [&ptr, origin = spec.origin](const TileElementIndex& index) {
+    return ptr(index + (origin - TileElementIndex(0, 0)));
+  };
+  EXPECT_EQ(spec.size, tile.size());
+  CHECK_TILE_PTR(subtile_ptr, tile);
+}
+
+template <class F, class T>
+void checkFullTile(F&& ptr, T&& tile, TileElementSize size) {
+  EXPECT_EQ(size, tile.size());
+  CHECK_TILE_PTR(ptr, tile);
+}
+
+// TileFutureOrConstTileSharedFuture should be
+// either hpx::future<Tile<T, D>> or hpx::shared_future<Tile<const T, D>>
+template <class TileFutureOrConstTileSharedFuture>
+void checkValidNonReady(const std::vector<TileFutureOrConstTileSharedFuture>& subtiles) {
+  for (const auto& subtile : subtiles) {
+    EXPECT_TRUE(subtile.valid());
+    EXPECT_FALSE(subtile.is_ready());
+  }
+}
+
+// TileFutureOrConstTileSharedFuture should be
+// either hpx::future<Tile<T, D>> or hpx::shared_future<Tile<const T, D>>
+// TileFuture should be hpx::future<Tile<T, D>>
+template <class F, class TileFutureOrConstTileSharedFuture, class TileFuture>
+void checkReadyAndDependencyChain(F&& tile_ptr, std::vector<TileFutureOrConstTileSharedFuture>& subtiles,
+                                  const std::vector<SubTileSpec>& specs, std::size_t last_dep,
+                                  TileFuture& next_tile_f) {
+  ASSERT_EQ(subtiles.size(), specs.size());
+  ASSERT_GT(subtiles.size(), last_dep);
```

Comment:
If I got it correctly, `last_dep` should be either `[subtiles.size() - 1` or `subtiles.size()`, right?

Reply:
Nope, it can be any subtile or the complete tile: i.e. [0, .size())